pallets/gas/src/property_tests.rs

```diff
@@ -0,0 +1,212 @@
+            node_ids.push(root_node);
+
+            // Only root has a max balance
+            assert_ok!(Gas::create(H256::random(), root_node, max_balance));
+
+            // Nodes on which `consume` was called
+            let mut marked_consumed = BTreeSet::new();
+            // Nodes that were created with `split` procedure
+            let mut unspec_ref_nodes = BTreeSet::new();
+            // Nodes that were created with `split_with_value` procedure
+            let mut spec_ref_nodes = BTreeSet::new();
+            // Total spent amount with `spent` procedure
+            let mut spent = 0;
+
+            for action in actions {
+                match action {
+                    GasTreeAction::SplitWithValue(parent_idx, amount) => {
+                        let parent = node_ids.ring_get(parent_idx).copied().expect("before each iteration there is at least 1 element; qed");
+                        let child = H256::random();
+
+                        if Gas::split_with_value(parent, child, amount).is_ok() {
+                            spec_ref_nodes.insert(child);
+                            node_ids.push(child)
+                        }
+                    }
+                    GasTreeAction::Split(parent_idx) => {
+                        let parent = node_ids.ring_get(parent_idx).copied().expect("before each iteration there is at least 1 element; qed");
+                        let child = H256::random();
+
+                        if Gas::split(parent, child).is_ok() {
+                            unspec_ref_nodes.insert(child);
+                            node_ids.push(child);
+                        }
+                    }
+                    GasTreeAction::Spend(from, amount) => {
+                        let from = node_ids.ring_get(from).copied().expect("before each iteration there is at least 1 element; qed");
+                        let limit = Gas::get_limit(from).unwrap().unwrap();
+                        let res = Gas::spend(from, amount);
+
+                        if limit < amount {
+                            assert_eq!(res, Err(Error::<Test>::InsufficientBalance.into()));
+                        } else {
+                            assert_ok!(res);
+                            spent += amount;
+                        }
+                    }
+                    GasTreeAction::Consume(id) => {
+                        let consuming = node_ids.ring_get(id).copied().expect("before each iteration there is at least 1 element; qed");
+                        if let Ok(removed_nodes) = utils::consume_node(consuming) {
+                            marked_consumed.insert(consuming);
+                            // Update ids
+                            node_ids.retain(|id| !removed_nodes.contains_key(id));
+
+                            // Search operation in a set is faster, then in a vector
+                            let remaining_ids = BTreeSet::from_iter(node_ids.iter().copied());
```

Comment:
should we compare lens before and after conversion to make sure that there is no duplicates?

Reply:
good idea

Reply:
checked in [2c75495](https://github.com/gear-tech/gear/pull/1069/commits/2c7549515885adc743197fc2de84c4c1e64cb432)